pkg/sidecar_controller/snapshot_controller.go

```diff
@@ -0,0 +1,576 @@
+// and making multiple bindings that are indistinguishable, resulting in
+// potential data loss.
+//
+// This controller is designed to work in active-passive high availability
+// mode. It *could* work also in active-active HA mode, all the object
+// transitions are designed to cope with this, however performance could be
+// lower as these two active controllers will step on each other toes
+// frequently.
+//
+// This controller supports both dynamic snapshot creation and pre-bound snapshot.
+// In pre-bound mode, objects are created with pre-defined pointers: a VolumeSnapshot
+// points to a specific VolumeSnapshotContent and the VolumeSnapshotContent also
+// points back for this VolumeSnapshot.
+//
+// The dynamic snapshot creation is multi-step process: first controller triggers
+// snapshot creation though csi volume plugin which should return a snapshot after
+// it is created successfully (however, the snapshot might not be ready to use yet if
+// there is an uploading phase). The creationTimestamp will be updated according to
+// VolumeSnapshot, and then a VolumeSnapshotContent object is created to represent
+// this snapshot. After that, the controller will keep checking the snapshot status
+// though csi snapshot calls. When the snapshot is ready to use, the controller set
+// the status "Bound" to true to indicate the snapshot is bound and ready to use.
+// If the createtion failed for any reason, the Error status is set accordingly.
+// In alpha version, the controller not retry to create the snapshot after it failed.
+// In the future version, a retry policy will be added.
+
+const pvcKind = "PersistentVolumeClaim"
+const apiGroup = ""
+const snapshotKind = "VolumeSnapshot"
+const snapshotAPIGroup = crdv1.GroupName
+
+const controllerUpdateFailMsg = "snapshot controller failed to update"
+
+const IsDefaultSnapshotClassAnnotation = "snapshot.storage.kubernetes.io/is-default-class"
+
+// syncContent deals with one key off the queue.  It returns false when it's time to quit.
+func (ctrl *csiSnapshotSideCarController) syncContent(content *crdv1.VolumeSnapshotContent) error {
+	klog.V(5).Infof("synchronizing VolumeSnapshotContent[%s]", content.Name)
+
+	if metav1.HasAnnotation(content.ObjectMeta, utils.AnnShouldDelete) {
+		klog.V(5).Infof("syncContent[%s]: check ShouldDelete annotation on content.", content.Name)
+		switch content.Spec.DeletionPolicy {
+		case crdv1.VolumeSnapshotContentRetain:
+			klog.V(4).Infof("VolumeSnapshotContent[%s]: policy is Retain, nothing to do", content.Name)
+
+		case crdv1.VolumeSnapshotContentDelete:
+			klog.V(4).Infof("VolumeSnapshotContent[%s]: policy is Delete", content.Name)
+			ctrl.deleteSnapshotContent(content)
+			klog.V(5).Infof("syncContent: check if we should remove Finalizer for VolumeSnapshotContent[%s]", content.Name)
+			// It is a deletion candidate if DeletionTimestamp is not nil and
+			// VolumeSnapshotContentFinalizer is set.
+			if utils.IsContentDeletionCandidate(content) {
```

Comment:
this seems to be a bit weird to me, the controller execute a "deleteSnapshotContent" call, and then check "IsContentDeletionCandidate"
I understand the finalizer should only be removed after physical snapshot has been cleaned. maybe rename this function?

Reply:
Renamed to deleteCSISnapshot